tx-pool/src/process.rs

```diff
@@ -624,16 +630,33 @@ impl TxPoolService {
         let fetched_cache = self.fetch_txs_verify_cache(retain.iter()).await;
 
         {
-            let mut tx_pool = self.tx_pool.write().await;
-            _update_tx_pool_for_reorg(
-                &mut tx_pool,
-                &attached,
-                detached_proposal_id,
-                snapshot,
-                &self.callbacks,
-            );
+            let txs_opt = {
+                // This closure is used to limit the lifetime of mutable tx_pool.
+                let mut tx_pool = self.tx_pool.write().await;
+
+                let txs_opt = if hardfork_during_detach || hardfork_during_attach {
+                    // The tx_pool is locked, remove all caches if has any hardfork.
+                    self.txs_verify_cache.write().await.clear();
+                    Some(tx_pool.drain_all_transactions())
+                } else {
+                    None
+                };
+
+                _update_tx_pool_for_reorg(
+                    &mut tx_pool,
+                    &attached,
+                    detached_proposal_id,
+                    snapshot,
+                    &self.callbacks,
+                );
+                self.readd_dettached_tx(&mut tx_pool, retain, fetched_cache);
+
+                txs_opt
+            };
 
-            self.readd_dettached_tx(&mut tx_pool, retain, fetched_cache)
+            if let Some(txs) = txs_opt {
+                self.try_process_txs(txs).await;
+            }
```

Comment:
At earlier version, I used empty `Vec<_>` to present `None`. May be it's better?

Backup old transactions, reorg, re-process old transactions.

Reply:
I think this is trivial.

Since this PR is hard to get approvals, and I already got one now.
Please let me fix this in one commit with other suggestions above **in another PR** after this merged.